model/message_broker.go

```diff
@@ -0,0 +1,164 @@
+	// message entry is an int sequentially incremented for each message
+	// published to the topic.
+	ForeignID string `bson:"foreign_id"`
+
+	// ResumeToken and Error should be checked when watching a topic.
+	ResumeToken []byte `bson:"-"`
+	Err         error  `bson:"-"`
+}
+
+var (
+	messageEntryTopicKey     = bsonutil.MustHaveTag(MessageEntry{}, "Topic")
+	messageEntryForeignIDKey = bsonutil.MustHaveTag(MessageEntry{}, "ForeignID")
+)
+
+// PublishToTopic creates a new message entry with the given id in the
+// corresponding collection for the given topic. If the topic does not exist in
+// the global registry, an error is returned.
+func PublishToTopic(ctx context.Context, env cedar.Environment, name string, id string) error {
+	topic, err := GetTopic(name)
+	if err != nil {
+		return err
+	}
+
+	return publishToTopic(ctx, env, topic, id)
+}
+
+// WatchTopic returns a MessageEntry channel fed by a change stream on the
+// topic's collection. The resume token specifies where to start after with the
+// change stream. If the topic does not exist in the global registry, an error
+// is returned.
+func WatchTopic(ctx context.Context, env cedar.Environment, name string, resumeToken []byte) (chan MessageEntry, error) {
+	topic, err := GetTopic(name)
+	if err != nil {
+		return nil, err
+	}
+
+	return watchTopic(ctx, env, topic, resumeToken)
+}
+
+// publishToTopic creates a new message entry with the foreign ID in the
+// corresponding collection for the given topic.
+func publishToTopic(ctx context.Context, env cedar.Environment, topic Topic, foreignID string) error {
+	entry := &MessageEntry{
+		Topic:     topic.Name(),
+		ForeignID: foreignID,
+	}
+	insertResult, err := env.GetDB().Collection(topicsCollection).InsertOne(ctx, entry)
+	grip.DebugWhen(err == nil, message.Fields{
+		"collection":   topic.Name(),
+		"insertResult": insertResult,
+		"op":           "publish to a topic",
+	})
+
+	return errors.Wrapf(err, "problem publishing to topic %s", topic.Name())
+}
+
+func watchTopic(ctx context.Context, env cedar.Environment, topic Topic, resumeToken []byte) (chan MessageEntry, error) {
+	opts := options.ChangeStream()
+	if len(resumeToken) > 0 {
+		opts = opts.SetResumeAfter(resumeToken)
+	}
+	cs, err := env.GetDB().Collection(topicsCollection).Watch(
+		ctx,
+		mongo.Pipeline{{{
+			Key: "$match", Value: bson.D{{
+				Key: messageEntryTopicKey, Value: topic.Name()},
+			},
+		}}},
+		opts,
+	)
+	if err != nil {
+		return nil, errors.Wrapf(err, "problem getting change streams for topic %s", topic.Name())
+	}
+
+	// TODO: figure out appropriate buffer size.
+	data := make(chan MessageEntry, 100)
+	go func() {
+		defer recovery.LogStackTraceAndContinue(fmt.Sprintf("watching topic %s", topic.Name()))
+		defer close(data)
+
+		for cs.Next(ctx) {
+			resumeToken = []byte(cs.ResumeToken())
+			entry := MessageEntry{}
+			entry.ResumeToken = resumeToken
+
+			if err = cs.Decode(&entry); err != nil {
```

Comment:
I don't understand how this works: I thought Decode would try to unmarshal into an event document.

Reply:
What do you mean by an event document?

Reply:
To me [the docs](https://docs.mongodb.com/manual/changeStreams/#open-a-change-stream) imply that iterating the change stream cursor returns [a change event document](https://docs.mongodb.com/manual/reference/change-events/#change-stream-output).

> To retrieve the data change event from the cursor, iterate the change stream cursor. For information on the change stream event, see Change Events.

Reply:
Oh right, yeah I would need to get the full document here.